### PR TITLE
[python] Fix training on subset constructed without params

### LINKED ISSUE #5170: lgb.train() encountering TypeError: 'NoneType' object is not iterable
## Description

I encountered this error when I tried to train with random forest type with the following parameters, 
```
num_boost_round=500

{'boosting_type': 'rf',
 'objective': 'regression',
 'bagging_freq': 1,
 'verbose': 1,
 'bagging_fraction': 0.5,
 'feature_fraction': 0.75,
 'min_data_in_leaf': 250}
```
The issue can be reproduced via the attached example which includes  fail_0.py and ml_rf.pkl.

I did search prior report wrt "TypeError: 'NoneType' object is not iterable" and found 2 (issue #1971 and #1994) but both seem to be related to the model fit() function instead of the train() function. In my case, the training data sets have no inf/nan. And in my use of the train(), I specified 3 columns as categorical (which appears to produce a user warning)

The console output upon running the reproducible script is as follows.

```console
$ python fail_0.py
C:\work\anaconda3\lib\site-packages\lightgbm\basic.py:2065: UserWarning: Using categorical_feature in Dataset.
  _log_warning('Using categorical_feature in Dataset.')
Traceback (most recent call last):
  File "fail_0.py", line 20, in <module>
    model = lgb.train(params=params,
  File "C:\work\anaconda3\lib\site-packages\lightgbm\engine.py", line 271, in train
    booster = Booster(params=params, train_set=train_set)
  File "C:\work\anaconda3\lib\site-packages\lightgbm\basic.py", line 2607, in __init__
    params.update(train_set.get_params())
TypeError: 'NoneType' object is not iterable
```

## Reproducible example

The script to reproduce the issue is very simple as shown below. It is in the attached tar.gz along with the pkl file 

```script
import _pickle as pickle
import pandas as pd
import numpy as np
import lightgbm as lgb

data_from_pickle_file = open("ml_rf.pkl", 'rb')
read_data_pickle = pickle.load(data_from_pickle_file)
data_from_pickle_file.close()

(params, num_boost_round, categoricals, outcome_data, label, train_idx, test_idx) = read_data_pickle

lgb_data = lgb.Dataset(data=outcome_data.drop(label, axis=1),
                       label=outcome_data[label],
                       categorical_feature=categoricals,
                       free_raw_data=False)

lgb_train = lgb_data.subset(train_idx.tolist()).construct()
lgb_test = lgb_data.subset(test_idx.tolist()).construct()

model = lgb.train(params=params,
                  train_set=lgb_train,
                  num_boost_round=num_boost_round)
```

## Environment info

LightGBM version or commit hash:  3.3.2
```
lightgbm                  3.3.2                    pypi_0    pypi
```

Command(s) you used to install LightGBM: 
```
pip install lightgbm
```

Shell:  git bash on Windows 10 

## Additional Comments

[fail.tar.gz](https://github.com/microsoft/LightGBM/files/8537342/fail.tar.gz)

### PR DESCRIPTION
**Pull request by StrikerRUS**
Fixed #5170.

Formally, we change the signature of the public `get_params()` method in this PR: it returns an empty dict instead of `None`. But I'm not sure about marking this PR a breaking change because previously this method could return empty dicts as well.

### CODE REVIEW

**Review by jmoralez — APPROVED** [review]
LGTM. After this change it's probably safe to change this function
https://github.com/microsoft/LightGBM/blob/2871b53e8e5e0ef23590410d7596d91173322ae4/python-package/lightgbm/basic.py#L292-L295 to remove the case for `None`.

**Review by jameslamb — APPROVED** [review]
Looks good, thanks!

I left one very very minor suggestion, but marking "approve" so you can merge immediately after addressing it if you agree.

**Comment by StrikerRUS** [comment]
@jmoralez 
> After this change it's probably safe to change this function

Hmm, are you sure in this?
`param_dict_to_str()` is used to convert parameters in many places of the codebase. While in most cases before those lines we have something like this
https://github.com/microsoft/LightGBM/blob/7d89ab403df9d524b373eac0674beaac658f02ab/python-package/lightgbm/basic.py#L2575
there is no guarantee we haven't forgotten to do such `None -> {}` replacement somewhere. I think it's better to leave that explicit `None` handling in `param_dict_to_str()`.

File: python-package/lightgbm/basic.py [hunk 1]
```diff
@@ -1353,7 +1353,7 @@ def get_params(self):
 
         Returns
         -------
-        params : dict or None
+        params : dict
```
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply python-package/lightgbm/basic.py:1356]: Addressed in 4b5fb51e9aafa155f18d34da9a31f34e678e1df2. Thank you for this suggestion!

[the remark this answers is not in the document]
Comment (jameslamb) [reply python-package/lightgbm/basic.py:1356]: thanks!

**Pull request merged**